pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/httproute_reconciler.go

```diff
@@ -0,0 +1,804 @@
+			return err
+		}
+		if !semanticHttpRouteEquals(desired, existing) {
+			if err := r.client.Update(ctx, desired); err != nil {
+				log.Error(err, "Failed to update explainer HttpRoute", "name", desired.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func (r *RawHTTPRouteReconciler) reconcileTopLevelHTTPRoute(ctx context.Context, isvc *v1beta1.InferenceService) error {
+	desired, err := createRawTopLevelHTTPRoute(isvc, r.ingressConfig, r.client)
+	if err != nil {
+		return err
+	}
+	if desired == nil {
+		return nil
+	}
+	if err := controllerutil.SetControllerReference(isvc, desired, r.scheme); err != nil {
+		log.Error(err, "Failed to set controller reference for top level HttpRoute", "name", isvc.Name)
+		return err
+	}
+	// reconcile httpRoute
+	existingHttpRoute := &gatewayapiv1.HTTPRoute{}
+	err = r.client.Get(ctx, types.NamespacedName{
+		Namespace: isvc.Namespace,
+		Name:      isvc.Name,
+	}, existingHttpRoute)
+	if err != nil {
+		if apierr.IsNotFound(err) {
+			log.Info("Creating top level HttpRoute resource", "name", isvc.Name)
+			if err := r.client.Create(ctx, desired); err != nil {
+				log.Error(err, "Failed to create top level HttpRoute", "name", desired.Name)
+				return err
+			}
+		} else {
+			return err
+		}
+	} else {
+		// Set ResourceVersion which is required for update operation.
+		desired.ResourceVersion = existingHttpRoute.ResourceVersion
+		// Do a dry-run update to avoid diffs generated by default values.
+		// This will populate our local httpRoute with any default values that are present on the remote version.
+		if err := r.client.Update(ctx, desired, client.DryRunAll); err != nil {
+			log.Error(err, "Failed to perform dry-run update on top level httpRoute", "name", desired.Name)
+			return err
+		}
+		if !semanticHttpRouteEquals(desired, existingHttpRoute) {
+			if err = r.client.Update(ctx, desired); err != nil {
+				log.Error(err, "Failed to update toplevel HttpRoute", "name", isvc.Name)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// ReconcileHTTPRoute reconciles the HTTPRoute resource
+func (r *RawHTTPRouteReconciler) Reconcile(ctx context.Context, isvc *v1beta1.InferenceService) error {
+	var err error
+	isInternal := false
+	// disable ingress creation if service is labelled with cluster local or kserve domain is cluster local
+	if val, ok := isvc.Labels[constants.NetworkVisibility]; ok && val == constants.ClusterLocalVisibility {
+		isInternal = true
+	}
+	if r.ingressConfig.IngressDomain == constants.ClusterLocalDomain {
+		isInternal = true
+	}
+	if !isInternal && !r.ingressConfig.DisableIngressCreation {
+		if err := r.reconcilePredictorHTTPRoute(ctx, isvc); err != nil {
+			return err
+		}
+		if isvc.Spec.Transformer != nil {
+			if err := r.reconcileTransformerHTTPRoute(ctx, isvc); err != nil {
+				return err
+			}
+		}
+		if isvc.Spec.Explainer != nil {
+			if err := r.reconcileExplainerHTTPRoute(ctx, isvc); err != nil {
+				return err
+			}
+		}
```

Comment:
so we are going to create individual http routes for predictor, transformer and explainer with the corresponding component host names, the top level http route directs to the component based on the verb: predict or explain, is this correct?

Reply:
Yes!